File: JavaScript/Daily/2024/Jan/Q872-LeafSimilarTrees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
    const tree1 = [];
    const tree2 = [];

    function traverse(node, tree) {
        if (!node) return;

        if (!node.left && !node.right) tree.push(node.val);
        traverse(node.left, tree);
        traverse(node.right, tree);
    }
    traverse(root1, tree1);
    traverse(root2, tree2);

    return JSON.stringify(tree1) === JSON.stringify(tree2);
};

File: Python/Leetcode75/Easy/BinaryTree/Q872-LeafSimilarTrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        nodes1 = []
        nodes2 = []

        self.dfs(root1, nodes1)
        self.dfs(root2, nodes2)

        return nodes1 == nodes2

    def dfs(self, node, leafs):
        if node == None:
            return
        if node.left:
            self.dfs(node.left, leafs)
        if node.right:
            self.dfs(node.right, leafs)
        if node.left == None and node.right == None:
            leafs.append(node.val)

    def leafSimilarOptimized(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return []
            if not node.left and not node.right:
                return [node.val]
            return dfs(node.left) + dfs(node.right)

        return dfs(root1) == dfs(root2)